File: 2022/day-5.rb
input = File.read('day-5-input.txt').split("\n")

stacks = [
  %w[N R G P],
  %w[J T B L F G D C],
  %w[M S V],
  %w[L S R C Z P],
  %w[P S L V C W D Q],
  %w[C T N W D M S],
  %w[H D G W P],
  %w[Z L P H S C M V],
  %w[R P F L W G Z],
]

data = input.grep /\Amove/

data.each do |l|
  l
    .match(/move (?<quantity>\d+) from (?<source>\d+) to (?<dest>\d+)/)
    .named_captures
    .transform_keys(&:to_sym)
    .transform_values(&:to_i) => { quantity: quantity, dest: dest, source: source }

  quantity.times do
    stacks[dest-1].push stacks[source-1].pop
  end
end

puts "Part 1: #{ stacks.map(&:last).join }"


stacks = [
  %w[N R G P],
  %w[J T B L F G D C],
  %w[M S V],
  %w[L S R C Z P],
  %w[P S L V C W D Q],
  %w[C T N W D M S],
  %w[H D G W P],
  %w[Z L P H S C M V],
  %w[R P F L W G Z],
]

data.each do |l|
  l
    .match(/move (?<quantity>\d+) from (?<source>\d+) to (?<dest>\d+)/)
    .named_captures
    .transform_keys(&:to_sym)
    .transform_values(&:to_i) => { quantity: quantity, dest: dest, source: source }

  s = []
  quantity.times do
    s.unshift stacks[source-1].pop
  end

  quantity.times do
    stacks[dest-1].push s.shift
  end
end

puts "Part 2: #{ stacks.map(&:last).join }"

File: 2022/day-5.ex
input = File.read!("day-5-input.txt")

stacks = [
  ~c[NRGP],
  ~c[JTBLFGDC],
  ~c[MSV],
  ~c[LSRCZP],
  ~c[PSLVCWDQ],
  ~c[CTNWDMS],
  ~c[HDGWP],
  ~c[ZLPHSCMV],
  ~c[RPFLWGZ]
]

defmodule AoC do
  def parseLine(line) do
    %{"quantity" => qty, "source" => source, "dest" => dest} =
      Regex.named_captures(
        ~r/move (?<quantity>\d+) from (?<source>\d+) to (?<dest>\d+)/,
        line
      )

    [qty, source, dest] |> Enum.map(&String.to_integer/1)
  end

  def move(stacks, quantity, source, dest) do
    stacks
    # pop
    # push
  end
end

data =
  input
  |> String.trim()
  |> String.split("\n")
  |> Enum.filter(fn l -> String.starts_with?(l, "move") end)
  |> Enum.map(&AoC.parseLine/1)

IO.inspect(data)

# result1 =
#   data
#   |> Enum.filter(fn pairs -> AoC.cover(pairs) || AoC.cover(Enum.reverse(pairs)) end)
#   |> Enum.count()

# IO.inspect(result1)
# IO.puts("Part1: #{result1}")

# result2 =
#   data
#   |> Enum.filter(&AoC.overlap/1)
#   |> Enum.count()

# IO.puts("Part2: #{result2}")
